pkg/internalapis/csinodetopology/csinodetopology.go

```diff
@@ -16,9 +16,240 @@ limitations under the License.
 
 package csinodetopology
 
+import (
+	"context"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	csinodetopologyv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/internalapis/csinodetopology/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
 const (
 	// CRDSingular represents the singular name of csinodetopology CRD.
 	CRDSingular = "csinodetopology"
 	// CRDPlural represents the plural name of csinodetopology CRD.
 	CRDPlural = "csinodetopologies"
+	// defaultTopologyCRWatcherTimeoutInMin is the default duration for which
+	// the topology service client will watch on the CSINodeTopology instance to check
+	// if the Status has been updated successfully.
+	defaultTopologyCRWatcherTimeoutInMin = 1
+	// maxTopologyCRWatcherTimeoutInMin is the maximum duration for which
+	// the topology service client will watch on the CSINodeTopology instance to check
+	// if the Status has been updated successfully.
+	maxTopologyCRWatcherTimeoutInMin = 2
+)
+
+// NodeInfo contains the information required for the TopologyService to
+// identify a node and retrieve its topology information.
+type NodeInfo struct {
+	// NodeName uniquely identifies the node in kubernetes.
+	NodeName string
+	// NodeID is a unique identifier of the NodeVM in vSphere.
+	NodeID string
+}
+
+// TopologyService is an interface which exposes functionality related to topology feature.
+type TopologyService interface {
+	// GetNodeTopologyLabels fetches the topology labels on the ancestors of NodeVM given the NodeInfo.
+	GetNodeTopologyLabels(ctx context.Context, info *NodeInfo) (map[string]string, error)
+}
+
+// volumeTopology stores information required for the TopologyService interface.
+type volumeTopology struct {
+	// k8sClient helps operate on CSINodeTopology custom resource.
+	k8sClient client.Client
+	// csiNodeTopologyWatcher is a watcher instance on the CSINodeTopology custom resource.
+	csiNodeTopologyWatcher *cache.ListWatch
+}
+
+var (
+	// volumeTopologyInstance is instance of volumeTopology and implements TopologyService interface.
+	volumeTopologyInstance *volumeTopology
+	// volumeTopologyInstanceLock is used for handling race conditions during read, write on volumeTopologyInstance
+	volumeTopologyInstanceLock = &sync.RWMutex{}
 )
+
+// InitTopologyServiceInterface returns a singleton implementation of the
+// TopologyService interface.
+func InitTopologyServiceInterface(ctx context.Context) (TopologyService, error) {
+	log := logger.GetLogger(ctx)
+
+	volumeTopologyInstanceLock.RLock()
```

Comment:
What's the need for the read lock? Why not just lock once. Since `NodeGetInfo` is only called once per node, this is a fairly inexpensive and infrequent operation.

Reply:
Agree that NodeGetInfo gets called only once but in the worst case, if NodeGetInfo errors out, it gets called repeatedly. These calls will be sequential, so should not be a problem if I remove RLock.
I am also thinking of cases where CreateVolume (the next PR) will call this service to figure out which NodeVMs belong to the given topology. In such cases, RLock might be useful.
Again, I got this idea from Divyen's code - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/apis/migration/migration.go#L107